x-pack/plugins/osquery/server/lib/telemetry/sender.ts

```diff
@@ -0,0 +1,298 @@
+          task.register(taskManager);
+          return task;
+        }
+      );
+    }
+  }
+
+  public getClusterID(): string | undefined {
+    return this.receiver?.getClusterInfo()?.cluster_uuid;
+  }
+
+  public start(
+    telemetryStart?: TelemetryPluginStart,
+    taskManager?: TaskManagerStartContract,
+    receiver?: TelemetryReceiver
+  ) {
+    this.telemetryStart = telemetryStart;
+    this.receiver = receiver;
+
+    if (taskManager && this.telemetryTasks) {
+      this.logger.debug(`Starting osquery telemetry tasks`);
+      this.telemetryTasks.forEach((task) => task.start(taskManager));
+    }
+
+    this.logger.debug(`Starting local task`);
+    setTimeout(() => {
+      this.sendIfDue();
+      this.intervalId = setInterval(() => this.sendIfDue(), this.checkIntervalMs);
+    }, this.initialCheckDelayMs);
+  }
+
+  public stop() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  public queueTelemetryEvents(events: TelemetryEvent[]) {
+    const qlength = this.queue.length;
+
+    if (events.length === 0) {
+      return;
+    }
+
+    this.logger.debug(`Queue events`);
+
+    if (qlength >= this.maxQueueSize) {
+      // we're full already
+      return;
+    }
+
+    if (events.length > this.maxQueueSize - qlength) {
+      this.telemetryUsageCounter?.incrementCounter({
+        counterName: createUsageCounterLabel(usageLabelPrefix.concat(['queue_stats'])),
+        counterType: 'docs_lost',
+        incrementBy: events.length,
+      });
+      this.telemetryUsageCounter?.incrementCounter({
+        counterName: createUsageCounterLabel(usageLabelPrefix.concat(['queue_stats'])),
+        counterType: 'num_capacity_exceeded',
+        incrementBy: 1,
+      });
+      this.queue.push(...this.processEvents(events.slice(0, this.maxQueueSize - qlength)));
+    } else {
+      this.queue.push(...this.processEvents(events));
+    }
+  }
+
+  public async isTelemetryOptedIn() {
+    this.isOptedIn = await this.telemetryStart?.getIsOptedIn();
+    return this.isOptedIn === true;
+  }
+
+  private async sendIfDue() {
+    if (this.isSending) {
+      return;
+    }
+
+    if (this.queue.length === 0) {
+      return;
+    }
+
+    try {
+      this.isSending = true;
+
+      this.isOptedIn = await this.isTelemetryOptedIn();
+      if (!this.isOptedIn) {
+        this.logger.debug(`Telemetry is not opted-in.`);
+        this.queue = [];
+        this.isSending = false;
+        return;
+      }
+
+      const clusterInfo = this.receiver?.getClusterInfo();
+
+      const [telemetryUrl, licenseInfo] = await Promise.all([
+        this.fetchTelemetryUrl('alerts-endpoint'),
+        this.receiver?.fetchLicenseInfo(),
+      ]);
+
+      this.logger.debug(`Telemetry URL: ${telemetryUrl}`);
+      this.logger.debug(
+        `cluster_uuid: ${clusterInfo?.cluster_uuid} cluster_name: ${clusterInfo?.cluster_name}`
+      );
+
+      const toSend: TelemetryEvent[] = cloneDeep(this.queue).map((event) => ({
```

Comment:
Just adding some benchmark notes here: https://www.measurethat.net/Benchmarks/Show/17003/0/lodash-clonedeep-vs-lodash-clone-vs-arraysplice-vs-obje

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/5469006/152127947-6d9a7f60-b071-431f-9790-f84e5b71d263.png">

We are trying to fight event loop delays as much as possible in Kibana, and this could be a candidate. Also note that we may need to change it in the other 2 implementations of this `sender.ts` across Kibana.

cc @elastic/kibana-core (more specifically @mshustov as he is in the performance working group)

Reply:
@afharo it's not entirely correct comparing `cloneDeep` and `Array. slice` - the former makes a deep copy of an array, the latter a shallow copy. Of course, `cloneDeep` is slower as it does much more work under the hood.
Looking at the code, I don't think we need `slice` either. `map` on the same line creates a new copy of an array.